exporter/awsemfexporter/datapoint.go

```diff
@@ -0,0 +1,326 @@
+	pdata.DoubleDataPointSlice
+}
+
+// DoubleHistogramDataPointSlice is a wrapper for pdata.DoubleHistogramDataPointSlice
+type DoubleHistogramDataPointSlice struct {
+	instrumentationLibraryName string
+	pdata.DoubleHistogramDataPointSlice
+}
+
+// DoubleSummaryDataPointSlice is a wrapper for pdata.DoubleSummaryDataPointSlice
+type DoubleSummaryDataPointSlice struct {
+	instrumentationLibraryName string
+	pdata.DoubleSummaryDataPointSlice
+}
+
+// At retrieves the IntDataPoint at the given index and performs rate calculation if necessary.
+func (dps IntDataPointSlice) At(i int) DataPoint {
+	metric := dps.IntDataPointSlice.At(i)
+	labels := createLabels(metric.LabelsMap(), dps.instrumentationLibraryName)
+	timestamp := unixNanoToMilliseconds(metric.Timestamp())
+
+	var metricVal interface{}
+	metricVal = metric.Value()
+	if dps.needsCalculateRate {
+		rateKey := createMetricKey(labels, dps.rateKeyParams)
+		rateTS := dps.timestamp
+		if timestamp > 0 {
+			// Use metric timestamp if available
+			rateTS = timestamp
+		}
+		metricVal = calculateRate(rateKey, metricVal, rateTS)
+	}
+
+	return DataPoint{
+		Value:     metricVal,
+		Labels:    labels,
+		Timestamp: timestamp,
+	}
+}
+
+// At retrieves the DoubleDataPoint at the given index and performs rate calculation if necessary.
+func (dps DoubleDataPointSlice) At(i int) DataPoint {
+	metric := dps.DoubleDataPointSlice.At(i)
+	labels := createLabels(metric.LabelsMap(), dps.instrumentationLibraryName)
+	timestamp := unixNanoToMilliseconds(metric.Timestamp())
+
+	var metricVal interface{}
+	metricVal = metric.Value()
+	if dps.needsCalculateRate {
+		rateKey := createMetricKey(labels, dps.rateKeyParams)
+		rateTS := dps.timestamp
+		if timestamp > 0 {
+			// Use metric timestamp if available
+			rateTS = timestamp
+		}
+		metricVal = calculateRate(rateKey, metricVal, rateTS)
+	}
+
+	return DataPoint{
+		Value:     metricVal,
+		Labels:    labels,
+		Timestamp: timestamp,
+	}
+}
+
+// At retrieves the DoubleHistogramDataPoint at the given index.
+func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
+	metric := dps.DoubleHistogramDataPointSlice.At(i)
+	labels := createLabels(metric.LabelsMap(), dps.instrumentationLibraryName)
+	timestamp := unixNanoToMilliseconds(metric.Timestamp())
+
+	var minBound, maxBound float64
+	bucketBounds := metric.ExplicitBounds()
+	if len(bucketBounds) > 0 {
+		minBound = bucketBounds[0]
+		maxBound = bucketBounds[len(bucketBounds)-1]
+	}
+
+	return DataPoint{
+		Value: &CWMetricStats{
+			Min:   minBound,
+			Max:   maxBound,
+			Count: metric.Count(),
+			Sum:   metric.Sum(),
+		},
+		Labels:    labels,
+		Timestamp: timestamp,
+	}
+}
+
+// At retrieves the DoubleSummaryDataPoint at the given index.
+func (dps DoubleSummaryDataPointSlice) At(i int) DataPoint {
+	metric := dps.DoubleSummaryDataPointSlice.At(i)
+	labels := createLabels(metric.LabelsMap(), dps.instrumentationLibraryName)
+	timestamp := unixNanoToMilliseconds(metric.Timestamp())
```

Comment:
`metric.Timestamp()` could be nil for some scenarios. Eg, The metrics generated from ECSMetadataMetrics receiver.

We may want to check if it is nil, we should use the current timestamp instead.

Reply:
We do that check in 
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1891/files/9cd18ee5346e703ce8af2150bdeff832da1dd26c?file-filters%5B%5D=.go#diff-14a056f0957da507e84dc694b704b81218c7d72a7454ce2a2d00354f7c725cc0R66